miner/block_builder.go

```diff
@@ -284,29 +263,44 @@ func (t *BlockBuilder) listenForTx() {
 				// not accepting txs when not synced
 				continue
 			}
-			if data != nil {
-
-				x, err := types.BytesAsSignedTransaction(data.Bytes())
-				if err != nil {
-					t.Log.Error("cannot parse incoming TX")
-					continue
-				}
-
-				id := types.GetTransactionId(x)
-				fullTx, err := t.txValidator.GetValidAddressableTx(x)
-				if err != nil {
-					t.Log.Error("Transaction validation failed for id=%v, err=%v", id, err)
-					continue
-				}
+			if data == nil {
+				continue
+			}
 
-				t.Log.With().Info("got new tx", log.TxId(hex.EncodeToString(id[:util.Min(5, len(id))])))
-				data.ReportValidation(IncomingTxProtocol)
-				t.TransactionPool.Put(types.GetTransactionId(x), fullTx)
+			tx, err := types.BytesAsTransaction(data.Bytes())
+			if err != nil {
+				t.With().Error("cannot parse incoming TX", log.Err(err))
+				continue
+			}
+			if err := tx.CalcAndSetOrigin(); err != nil {
+				t.With().Error("failed to calc transaction origin", log.TxId(tx.Id().ShortString()), log.Err(err))
+				continue
+			}
+			if !t.txValidator.AddressExists(tx.Origin()) {
+				t.With().Error("transaction origin does not exist", log.String("transaction", tx.String()),
+					log.TxId(tx.Id().ShortString()), log.String("origin", tx.Origin().Short()), log.Err(err))
+				continue
+			}
+			if err := t.txValidator.ValidateNonceAndBalance(tx); err != nil {
+				t.With().Error("nonce and balance validation failed", log.TxId(tx.Id().ShortString()), log.Err(err))
+				continue
 			}
+			t.Log.With().Info("got new tx", log.TxId(tx.Id().ShortString()))
+			data.ReportValidation(IncomingTxProtocol)
+			t.TransactionPool.Put(tx.Id(), tx)
 		}
 	}
 }
 
+func (t *BlockBuilder) ValidateAndAddTxToPool(tx *types.Transaction) error {
```

Comment:
please separate the writing of the tx to the pool from validating it, or change the name of the func since it's misleading